src/Castle.Windsor/MicroKernel/Releasers/LifecycledComponentsReleasePolicy.cs

```diff
@@ -84,33 +84,43 @@ private Burden[] TrackedObjects
 		{
 			get
 			{
-				using (var holder = @lock.ForReading(false))
+				@lock.EnterReadLock();
+
+				try
 				{
-					if (holder.LockAcquired == false)
-					{
-						// TODO: that's sad... perhaps we should have waited...? But what do we do now? We're in the debugger. If some thread is keeping the lock
-						// we could wait indefinatelly. I guess the best way to proceed is to add a 200ms timepout to accquire the lock, and if not succeeded
-						// assume that the other thread just waits and is not going anywhere and go ahead and read this anyway...
-					}
 					var array = instance2Burden.Values.ToArray();
 					return array;
 				}
+				finally
+				{
+					@lock.ExitReadLock();
+				}
 			}
 		}
 
 		public void Dispose()
 		{
 			KeyValuePair<object, Burden>[] burdens;
-			using (@lock.ForWriting())
+
+			@lock.EnterWriteLock();
+
+			try
 			{
+
 				if (trackedComponentsDiagnostic != null)
 				{
 					trackedComponentsDiagnostic.TrackedInstancesRequested -= trackedComponentsDiagnostic_TrackedInstancesRequested;
 					trackedComponentsDiagnostic = null;
 				}
+
 				burdens = instance2Burden.ToArray();
 				instance2Burden.Clear();
 			}
+			finally
+			{
+				@lock.ExitWriteLock();
+			}
+
```

Comment:
A note about the `foreach` loop following this code. It calls `Release` on each `burden` in the `burdens` collection. Each `Release` will take a write lock. That seems kind of inefficient to me. If I've read this correctly, then this perhaps could be more efficient, something along the lines of:

1. Move the actual logic of `Release` (all except the lock taking) into a new, separate method `ReleaseWithoutTakingLock`.
2. Delegate to that new method from inside `Release`.
3. Before the `foreach` loop (here), take a lock. Inside the loop, use the new non-locking method.

I could be wrong, as I haven't studied this code in detail. But perhaps worth giving a thought or two.